File: server/storage.ts
import {
   appointments,
  type Appointment,
  type InsertAppointment,
  type ChatMessage,
  type InsertChatMessage,
  type Conversation,
  type InsertConversation,
  type Doctor,
  type ConversationSlots,
} from "../shared/schema";

import { db } from "./db";
import { eq } from "drizzle-orm";
import { randomUUID } from "crypto";
import fs from "fs";
import path from "path";

export interface IStorage {
  createAppointment(appointment: InsertAppointment): Promise<Appointment>;
  getAppointment(id: string): Promise<Appointment | undefined>;
  getAllAppointments(): Promise<Appointment[]>;

  createChatMessage(message: InsertChatMessage): Promise<ChatMessage>;
  getMessagesByConversation(conversationId: string): Promise<ChatMessage[]>;

  createConversation(conversation: InsertConversation): Promise<Conversation>;
  getConversation(id: string): Promise<Conversation | undefined>;
  getAllConversations(): Promise<Conversation[]>;
  updateConversationSlots(id: string, slots: ConversationSlots): Promise<void>;
  updateConversationTitle(id: string, title: string): Promise<void>;

  getDoctors(): Promise<Doctor[]>;
}

function parseCSVLine(line: string): string[] {
  const values: string[] = [];
  let currentValue = '';
  let insideQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    const nextChar = line[i + 1];

    if (char === '"') {
      if (insideQuotes && nextChar === '"') {
        currentValue += '"';
        i++;
      } else {
        insideQuotes = !insideQuotes;
      }
    } else if (char === ',' && !insideQuotes) {
      values.push(currentValue.trim());
      currentValue = '';
    } else {
      currentValue += char;
    }
  }
  
  if (currentValue) {
    values.push(currentValue.trim());
  }
  
  return values;
}

function parseCSV(filePath: string): Doctor[] {
  const fileContent = fs.readFileSync(filePath, 'utf-8');
  const lines = fileContent.split('\n');
  const headers = parseCSVLine(lines[0]);
  
  const doctors: Doctor[] = [];
  
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    
    const values = parseCSVLine(line);
    const row: Record<string, string> = {};
    
    headers.forEach((header, index) => {
      row[header] = values[index] || '';
    });
    
    doctors.push({
      id: row['doctor_id'] || `dr-${i}`,
      name: row['doctor_name_en'] || 'Unknown',
      specialty: row['specialty'] || 'General Practice',
      subSpecialty: row['sub_specialty'] || undefined,
      hospitalName: row['hospital_name'] || undefined,
      locationArea: row['location_area'] || undefined,
      rating: row['rating'] ? parseFloat(row['rating']) : undefined,
      fee: row['fee'] ? parseInt(row['fee']) : undefined,
      avatar: '',
      available: row['telemedicine'] === 'True',
    });
  }
  
  return doctors;
}

// function parseCSV(filePath: string): Doctor[] {
//   const fileContent = fs.readFileSync(filePath, 'utf-8');
//   const lines = fileContent.split('\n');
//   const headers = lines[0].split(',').map(h => h.trim());
  
//   const doctors: Doctor[] = [];
  
//   for (let i = 1; i < lines.length; i++) {
//     const line = lines[i].trim();
//     if (!line) continue;
    
//     const values = line.split(',').map(v => v.trim());
//     const row: Record<string, string> = {};
    
//     headers.forEach((header, index) => {
//       row[header] = values[index] || '';
//     });
    
//     doctors.push({
//       id: row['doctor_id'] || `dr-${i}`,
//       name: row['doctor_name_en'] || 'Unknown',
//       specialty: row['specialty'] || 'General Practice',
//       avatar: '',
//       available: row['telemedicine'] === 'True' || Math.random() > 0.3,
//     });
//   }
  
//   return doctors;
// }

export class MemStorage implements IStorage {
  private appointments: Map<string, Appointment>;
  private chatMessages: Map<string, ChatMessage>;
  private conversations: Map<string, Conversation>;
  private doctors: Doctor[];

  constructor() {
    this.appointments = new Map();
    this.chatMessages = new Map();
    this.conversations = new Map();

    try {
      const csvPath = path.resolve(import.meta.dirname, '../attached_assets/doctors_300_dataset_1763920263351.csv');
      this.doctors = parseCSV(csvPath);
    } catch (error) {
      console.warn('Failed to load doctors from CSV, using fallback', error);
      this.doctors = [
        {
          id: "dr-sarah-johnson",
          name: "Dr. Sarah Johnson",
          specialty: "Endocrinologist",
          avatar: "",
          available: true,
        },
        {
          id: "dr-michael-chen",
          name: "Dr. Michael Chen",
          specialty: "Cardiologist",
          avatar: "",
          available: true,
        },
        {
          id: "dr-emily-rodriguez",
          name: "Dr. Emily Rodriguez",
          specialty: "Oncologist",
          avatar: "",
          available: true,
        },
      ];
    }
  }

  async createAppointment(
    insertAppointment: InsertAppointment,
  ): Promise<Appointment> {
    const id = randomUUID();
    const appointment: Appointment = {
      ...insertAppointment,
      id,
      appointmentDate: new Date(insertAppointment.appointmentDate),
      status: "confirmed",
      createdAt: new Date(),

      // FIX: ensure reason is never undefined
      reason: insertAppointment.reason ?? null,
    };
    this.appointments.set(id, appointment);
    return appointment;
  }

  async getAppointment(id: string): Promise<Appointment | undefined> {
    return this.appointments.get(id);
  }

  async getAllAppointments(): Promise<Appointment[]> {
    return Array.from(this.appointments.values());
  }

  async createChatMessage(
    insertMessage: InsertChatMessage,
  ): Promise<ChatMessage> {
    const id = randomUUID();
    const message: ChatMessage = {
      ...insertMessage,
      id,
      createdAt: new Date(),

      // FIX: normalize optional fields
      citations: insertMessage.citations ?? null,
      metadata: insertMessage.metadata ?? null,
    };
    this.chatMessages.set(id, message);
    return message;
  }

  async getMessagesByConversation(
    conversationId: string,
  ): Promise<ChatMessage[]> {
    return Array.from(this.chatMessages.values())
      .filter((msg) => msg.conversationId === conversationId)
      .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime());
  }

  async createConversation(
    insertConversation: InsertConversation,
  ): Promise<Conversation> {
    const id = randomUUID();
    const now = new Date();
    const conversation: Conversation = {
      ...insertConversation,
      id,
      title: insertConversation.title || null,
      slots: insertConversation.slots || null,
      createdAt: now,
      updatedAt: now,
    };
    this.conversations.set(id, conversation);
    return conversation;
  }

  async getConversation(id: string): Promise<Conversation | undefined> {
    return this.conversations.get(id);
  }

  async getAllConversations(): Promise<Conversation[]> {
    return Array.from(this.conversations.values()).sort(
      (a, b) => b.updatedAt.getTime() - a.updatedAt.getTime(),
    );
  }

  async updateConversationSlots(
    id: string,
    slots: ConversationSlots,
  ): Promise<void> {
    const conversation = this.conversations.get(id);
    if (conversation) {
      conversation.slots = JSON.stringify(slots);
      conversation.updatedAt = new Date();
    }
  }

  async updateConversationTitle(id: string, title: string): Promise<void> {
    const conversation = this.conversations.get(id);
    if (conversation) {
      conversation.title = title;
      conversation.updatedAt = new Date();
    }
  }

  async getDoctors(): Promise<Doctor[]> {
    return this.doctors;
  }
}

export class PostgresStorage implements IStorage {
  constructor(private fallback: MemStorage) {}

  // --- Appointments stored in PostgreSQL ---

  async createAppointment(appointment: InsertAppointment): Promise<Appointment> {
    // Ensure appointmentDate is a real Date object for Drizzle/Postgres
    const appointmentDate =
      appointment.appointmentDate instanceof Date
        ? appointment.appointmentDate
        : new Date(appointment.appointmentDate as any);

    if (isNaN(appointmentDate.getTime())) {
      throw new Error("Invalid appointmentDate value");
    }

    const [row] = await db
      .insert(appointments)
      .values({
        ...appointment,
        appointmentDate,
        status: (appointment as any).status ?? "confirmed",
      })
      .returning();

    return row;
  }


  async getAppointment(id: string): Promise<Appointment | undefined> {
    const [row] = await db
      .select()
      .from(appointments)
      .where(eq(appointments.id, id));

    return row ?? undefined;
  }

  async getAllAppointments(): Promise<Appointment[]> {
    const rows = await db.select().from(appointments);
    return rows;
  }

  // --- Everything else stays in-memory via the existing MemStorage ---

  async createChatMessage(msg: InsertChatMessage): Promise<ChatMessage> {
    return this.fallback.createChatMessage(msg);
  }

  async getMessagesByConversation(conversationId: string): Promise<ChatMessage[]> {
    return this.fallback.getMessagesByConversation(conversationId);
  }

  async createConversation(conv: InsertConversation): Promise<Conversation> {
    return this.fallback.createConversation(conv);
  }

  async getConversation(id: string): Promise<Conversation | undefined> {
    return this.fallback.getConversation(id);
  }

  async getAllConversations(): Promise<Conversation[]> {
    return this.fallback.getAllConversations();
  }

  async updateConversationSlots(id: string, slots: ConversationSlots): Promise<void> {
    return this.fallback.updateConversationSlots(id, slots);
  }

  async updateConversationTitle(id: string, title: string): Promise<void> {
    return this.fallback.updateConversationTitle(id, title);
  }

  async getDoctors(): Promise<Doctor[]> {
    return this.fallback.getDoctors();
  }
}

const baseStorage = new MemStorage();
export const storage = new PostgresStorage(baseStorage);

File: python-backend/storage.py
import uuid
import csv
import json
from datetime import datetime
from typing import List, Optional, Dict
from sqlalchemy import create_engine, Column, String, Integer, Text, DateTime
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from models import Appointment, Conversation, ChatMessage, Doctor
from config import DATABASE_URL

# ---------- Database setup ----------
engine = create_engine(DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
Base = declarative_base()

# ---------- SQLAlchemy model for uploaded chunks ----------
class UploadedChunkDB(Base):
    __tablename__ = "uploaded_chunks"
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    conversation_id = Column(String, index=True, nullable=False)
    filename = Column(String, nullable=False)
    chunk_index = Column(Integer, nullable=False)
    content = Column(Text, nullable=False)
    embedding = Column(Text, nullable=False)   # JSON string of list
    source = Column(String, nullable=False)
    created_at = Column(DateTime, default=datetime.now, nullable=False)

# ---------- In-memory fallback (if DB not configured) ----------
_memory_appointments: Dict[str, Appointment] = {}
_memory_conversations: Dict[str, Conversation] = {}
_memory_messages: Dict[str, ChatMessage] = {}
_memory_doctors: List[Doctor] = []

def init_db():
    """Create all tables if they don't exist"""
    Base.metadata.create_all(engine)

# ---------- Storage class with both memory and DB methods ----------
class Storage:
    # ----- Doctors -----
    @staticmethod
    async def get_doctors() -> List[Doctor]:
        if _memory_doctors:
            return _memory_doctors
        csv_path = "../attached_assets/doctors_300_dataset_1763920263351.csv"
        import os
        if os.path.exists(csv_path):
            doctors = []
            with open(csv_path, 'r', encoding='utf-8') as f:
                reader = csv.DictReader(f)
                for row in reader:
                    doctors.append(Doctor(
                        id=row.get('doctor_id', str(uuid.uuid4())),
                        name=row.get('doctor_name_en', 'Unknown'),
                        specialty=row.get('specialty', 'General Practice'),
                        subSpecialty=row.get('sub_specialty'),
                        hospitalName=row.get('hospital_name'),
                        locationArea=row.get('location_area'),
                        rating=float(row['rating']) if row.get('rating') else None,
                        fee=int(row['fee']) if row.get('fee') else None,
                        avatar="",
                        available=row.get('telemedicine') == 'True'
                    ))
            _memory_doctors.extend(doctors)
            return _memory_doctors
        # Fallback
        if not _memory_doctors:
            _memory_doctors.append(Doctor(id="dr-sarah-johnson", name="Dr. Sarah Johnson", specialty="Endocrinologist", avatar="", available=True))
            _memory_doctors.append(Doctor(id="dr-michael-chen", name="Dr. Michael Chen", specialty="Cardiologist", avatar="", available=True))
            _memory_doctors.append(Doctor(id="dr-emily-rodriguez", name="Dr. Emily Rodriguez", specialty="Oncologist", avatar="", available=True))
        return _memory_doctors

    # ----- Appointments -----
    @staticmethod
    async def create_appointment(data: dict) -> Appointment:
        appt_id = str(uuid.uuid4())
        now = datetime.now()
        appointment = Appointment(
            id=appt_id,
            patientName=data['patientName'],
            patientEmail=data['patientEmail'],
            doctorId=data['doctorId'],
            appointmentDate=data['appointmentDate'],
            timezone=data['timezone'],
            status="confirmed",
            reason=data.get('reason'),
            createdAt=now
        )
        _memory_appointments[appt_id] = appointment
        return appointment

    @staticmethod
    async def get_appointment(appt_id: str) -> Optional[Appointment]:
        return _memory_appointments.get(appt_id)

    @staticmethod
    async def get_all_appointments() -> List[Appointment]:
        return list(_memory_appointments.values())

    # ----- Conversations -----
    @staticmethod
    async def create_conversation() -> Conversation:
        conv_id = str(uuid.uuid4())
        now = datetime.now()
        conv = Conversation(id=conv_id, title=None, slots=None, createdAt=now, updatedAt=now)
        _memory_conversations[conv_id] = conv
        return conv

    @staticmethod
    async def get_conversation(conv_id: str) -> Optional[Conversation]:
        return _memory_conversations.get(conv_id)

    @staticmethod
    async def get_all_conversations() -> List[Conversation]:
        return sorted(_memory_conversations.values(), key=lambda c: c.updatedAt, reverse=True)

    @staticmethod
    async def update_conversation_slots(conv_id: str, slots: dict) -> None:
        conv = _memory_conversations.get(conv_id)
        if conv:
            conv.slots = json.dumps(slots)
            conv.updatedAt = datetime.now()

    @staticmethod
    async def update_conversation_title(conv_id: str, title: str) -> None:
        conv = _memory_conversations.get(conv_id)
        if conv:
            conv.title = title
            conv.updatedAt = datetime.now()

    # ----- Chat Messages -----
    @staticmethod
    async def create_chat_message(conv_id: str, role: str, content: str, citations: List[str] = None) -> ChatMessage:
        msg_id = str(uuid.uuid4())
        msg = ChatMessage(
            id=msg_id,
            conversationId=conv_id,
            role=role,
            content=content,
            citations=citations,
            createdAt=datetime.now()
        )
        _memory_messages[msg_id] = msg
        return msg

    @staticmethod
    async def get_messages_by_conversation(conv_id: str) -> List[ChatMessage]:
        return [m for m in _memory_messages.values() if m.conversationId == conv_id]

# ---------- Uploaded chunks DB operations ----------
async def save_uploaded_chunks(conversation_id: str, filename: str, chunks: list):
    """chunks: list of dicts with keys: content, embedding, source"""
    session = SessionLocal()
    try:
        for idx, chunk in enumerate(chunks):
            db_chunk = UploadedChunkDB(
                conversation_id=conversation_id,
                filename=filename,
                chunk_index=idx,
                content=chunk["content"],
                embedding=json.dumps(chunk["embedding"]),
                source=chunk["source"],
            )
            session.add(db_chunk)
        session.commit()
    except Exception as e:
        session.rollback()
        raise e
    finally:
        session.close()

async def get_uploaded_chunks(conversation_id: str):
    session = SessionLocal()
    try:
        rows = session.query(UploadedChunkDB).filter(
            UploadedChunkDB.conversation_id == conversation_id
        ).order_by(UploadedChunkDB.chunk_index).all()
        print(f"DEBUG: Found {len(rows)} rows in DB for conv {conversation_id}")
        chunks = []
        for row in rows:
            chunks.append({
                "source": row.source,
                "content": row.content,
                "embedding": json.loads(row.embedding)
            })
        return chunks
    finally:
        session.close()

async def delete_uploaded_chunks_for_conversation(conversation_id: str):
    session = SessionLocal()
    try:
        session.query(UploadedChunkDB).filter(
            UploadedChunkDB.conversation_id == conversation_id
        ).delete()
        session.commit()
    except Exception:
        session.rollback()
        raise
    finally:
        session.close()